bip-musig2/reference.py

```diff
@@ -389,420 +388,270 @@ def assertRaises(exception, try_fn, except_fn):
-        'DFF1D77F2A671C5F36183726DB2341BE58FEAE1DA2DECED843240F7B502BA659',
-        '3590A94E768F8E1815C2F24B4D80A8E3149316C3518CE7B7AD338368D038CA66',
-    ])
-
-    expected = fromhex_all([
-        'E5830140512195D74C8307E39637CBE5FB730EBEAB80EC514CF88A877CEEEE0B',
-        'D70CD69A2647F7390973DF48CBFA2CCC407B8B2D60B08C5F1641185C7998A290',
-        '81A8B093912C9E481408D09776CEFB48AEB8B65481B6BAAFB3C5810106717BEB',
-        '2EB18851887E7BDC5E830E89B19DDBC28078F1FA88AAD0AD01CA06FE4F80210B',
-    ])
-
-    # Vector 1
-    assert get_pk(key_agg([X[0], X[1], X[2]])) == expected[0]
-    # Vector 2
-    assert get_pk(key_agg([X[2], X[1], X[0]])) == expected[1]
-    # Vector 3
-    assert get_pk(key_agg([X[0], X[0], X[0]])) == expected[2]
-    # Vector 4
-    assert get_pk(key_agg([X[0], X[0], X[1], X[1]])) == expected[3]
-
-    # Vector 5: Invalid public key
-    invalid_pk = bytes.fromhex('0000000000000000000000000000000000000000000000000000000000000005')
-    assertRaises(InvalidContributionError,
-                 lambda: key_agg([X[0], invalid_pk]),
-                 lambda e: e.signer == 1 and e.contrib == "pubkey")
-    # Vector 6: Public key exceeds field size
-    invalid_pk = bytes.fromhex('FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC30')
-    assertRaises(InvalidContributionError,
-                 lambda: key_agg([X[0], invalid_pk]),
-                 lambda e: e.signer == 1 and e.contrib == "pubkey")
-    # Vector 7: Tweak is out of range
-    invalid_tweak = bytes.fromhex('FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141')
-    assertRaises(ValueError,
-                 lambda: key_agg_and_tweak([X[0], X[1]], [invalid_tweak], [True]),
-                 lambda e: str(e) == 'The tweak must be less than n.')
-    # Vector 8: Intermediate tweaking result is point at infinity
-    G_ = bytes_from_point(G)
-    coeff = bytes_from_int(n - key_agg_coeff([G_], G_))
-    assertRaises(ValueError,
-                 lambda: key_agg_and_tweak([G_], [coeff], [False]),
-                 lambda e: str(e) == 'The result of tweaking cannot be infinity.')
+    with open('key_agg_vectors.json') as f:
+        test_data = json.load(f)
+
+    X = fromhex_all(test_data["pubkeys"])
+    T = fromhex_all(test_data["tweaks"])
+    valid_test_cases = test_data["valid_test_cases"]
+    error_test_cases = test_data["error_test_cases"]
+
+    for test_case in valid_test_cases:
+        pubkeys = [X[i] for i in test_case["key_indices"]]
+        expected = bytes.fromhex(test_case["expected"])
+
+        assert get_pk(key_agg(pubkeys)) == expected
+
+    expected_errors = [
+        (InvalidContributionError, lambda e: e.signer == 1 and e.contrib == "pubkey"),
+        (InvalidContributionError, lambda e: e.signer == 1 and e.contrib == "pubkey"),
+        (ValueError, lambda e: str(e) == 'The tweak must be less than n.'),
+        (ValueError, lambda e: str(e) == 'The result of tweaking cannot be infinity.')
+    ]
```

Comment:
It would be good to be able to express `InvalidContribution(signer)` and `ValueError` in the JSON file. It may sound like overkill on first glance but these the difference between the error paths should can be critical for callers of some implementations, e.g., when they blame the wrong signer. 

With the current version, we now spread the specification of what is the right error across the JSON file and our Python file.

(On the other hand, fixing `contrib` is probably too much,  and fixing the error string is too much for sure.)

Reply:
Good point. Do you like any of these options:

- Add more details to "comment", e.g. "Invalid public key" -> "Invalid contribution from signer 1 (public key)"
- Add an optional "misbehaving_signer" field to the JSON

Reply:
> Add an optional "misbehaving_signer" field to the JSON

Having it machine readable was what I had in mind, though we need a way to distinguish `InvalidContribution(None)` (blame aggregator)  and `ValueError`. Maybe `ValueError` if the field not present (null) and some special value ("-1" or "aggregator") for `InvalidContribution(None)`? May then be more consistent to call the field `invalid_contribution` or `misbehaving_party` because it can include the aggregator. 


(Ok, "-1" looks like we're back in the 80s but well, this is test code...)

Reply:
It was actually easier for me to just do the overkill thing where I fix both `contrib` and the error message...

Reply:
looks great now, I'll do some final review/mutation testing once #24 is in